### PR TITLE
Fix for getting Wazuh versions >=3.10

### LINKED ISSUE #3347: Error getting version with Wazuh 3.10 or higher
Hi team,

When we install Wazuh and API with `3.10`, an error happens when version is checked:

```bash
# curl -u foo:bar -k -X GET "http://127.0.0.1:55000/agents/000/config/syscheck/internal?pretty"
{
   "error": 1000,
   "message": "Internal error"
}
```

After debugging this call:

```bash
WazuhAPI 2019-05-20 09:33:02 foo: CMD - Command: /var/ossec/framework/python/bin/python3 args:/var/ossec/api/models/wazuh-api.py stdin:{"function":"/agents/:agent_id/config/:component/:configuration","arguments":{"agent_id":"000","component":"syscheck","configuration":"internal","wait_for_complete":false},"ossec_path":"/var/ossec"}
WazuhAPI 2019-05-20 09:33:02 foo: CMD - Exit code: 0
WazuhAPI 2019-05-20 09:33:02 foo: CMD - STDOUT:
---
{"message": "Invalid version format.", "error": 1000}
```

An exception is raised if `minor` (or otherwise)  has more than one digit:

https://github.com/wazuh/wazuh/blob/4476fe8036a5665b2fef35ef8d5ff74804767bab/framework/wazuh/agent.py#L2459

https://github.com/wazuh/wazuh/blob/4476fe8036a5665b2fef35ef8d5ff74804767bab/framework/wazuh/utils.py#L518-L532

Regular expression `pattern` should be updated for giving support to new Wazuh versions.

Furthermore, it is necessary to review our code for checking if there are more errors as this.

Best regards,

Demetrio.

### PR DESCRIPTION
**Pull request by druizz90**
Hi team,

This PR closes #3347. I made a fix for getting Wazuh version properly. Numbers of version are casted to integers and regular expression to catch these numbers was updated for allowing two digits (or more) numbers.

API calls that were not working before the fix are OK now:

```bash
# curl -u foo:bar -k -X GET "http://127.0.0.1:55000/agents/000/config/syscheck/internal?pretty"
{
   "error": 0,
   "data": {
      "internal": {
         "syscheck": {
            "sleep": 1,
            "sleep_after": 100,
            "rt_delay": 10,
            "default_max_depth": 256,
            "symlink_scan_interval": 0,
            "debug": 0,
            "file_max_size": 1073741824,
            "max_audit_entries": 256
         },
         "rootcheck": {
            "sleep": 50
         }
      }
   }
}
```

```bash
# curl -u foo:bar -k -X GET "http://127.0.0.1:55000/agents/000/config/monitor/internal?pretty"   
{
   "error": 0,
   "data": {
      "monitord": {
         "day_wait": 10,
         "compress": 1,
         "sign": 1,
         "monitor_agents": 1,
         "keep_log_days": 31,
         "rotate_log": 1,
         "size_rotate": 536870912,
         "daily_rotations": 12,
         "delete_old_agents": 0
      }
   }
}

```

Best regards,

Demetrio

### CODE REVIEW

**Comment by druizz90** [comment]
## Tests performed

#### Unit tests

```bash
# /var/ossec/framework/python/bin/pytest -s framework/wazuh/tests/test_util.py               
==================================================== test session starts ====================================================
platform linux -- Python 3.7.2, pytest-4.5.0, py-1.8.0, pluggy-0.11.0
rootdir: /root/wazuh-old/wazuh/framework
collected 9 items                                                                                                           

framework/wazuh/tests/test_util.py .........

================================================= 9 passed in 0.08 seconds ==================================================
localhost:~/wazuh-old/wazuh (fix-wpk-versions-3353=) # /var/ossec/framework/python/bin/pytest -s framework/wazuh/tests/test_util.py -vv
==================================================== test session starts ====================================================
platform linux -- Python 3.7.2, pytest-4.5.0, py-1.8.0, pluggy-0.11.0 -- /var/ossec/framework/python/bin/python3.7
cachedir: .pytest_cache
rootdir: /root/wazuh-old/wazuh/framework
collected 9 items                                                                                                           

framework/wazuh/tests/test_util.py::test_version[Wazuh v3.5.0-Wazuh v3.5.2] PASSED
framework/wazuh/tests/test_util.py::test_version[Wazuh v3.6.1-Wazuh v3.6.3] PASSED
framework/wazuh/tests/test_util.py::test_version[Wazuh v3.7.2-Wazuh v3.8.0] PASSED
framework/wazuh/tests/test_util.py::test_version[Wazuh v3.8.0-Wazuh v3.8.1] PASSED
framework/wazuh/tests/test_util.py::test_version[Wazuh v3.9.0-Wazuh v3.9.2] PASSED
framework/wazuh/tests/test_util.py::test_version[Wazuh v3.9.10-Wazuh v3.9.14] PASSED
framework/wazuh/tests/test_util.py::test_version[Wazuh v3.10.1-Wazuh v3.10.10] PASSED
framework/wazuh/tests/test_util.py::test_version[Wazuh v4.10.10-Wazuh v4.11.0] PASSED
framework/wazuh/tests/test_util.py::test_version[Wazuh v5.1.15-Wazuh v5.2.0] PASSED

================================================= 9 passed in 0.08 seconds ==================================================
```

File: framework/wazuh/tests/test_util.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+@pytest.mark.parametrize('version1, version2', [
+    ('Wazuh v3.5.0', 'Wazuh v3.5.2'),
+    ('Wazuh v3.6.1', 'Wazuh v3.6.3'),
+    ('Wazuh v3.7.2', 'Wazuh v3.8.0'),
+    ('Wazuh v3.8.0', 'Wazuh v3.8.1'),
+    ('Wazuh v3.9.0', 'Wazuh v3.9.2'),
+    ('Wazuh v3.9.10', 'Wazuh v3.9.14'),
+    ('Wazuh v3.10.1', 'Wazuh v3.10.10'),
+    ('Wazuh v4.10.10', 'Wazuh v4.11.0'),
+    ('Wazuh v5.1.15', 'Wazuh v5.2.0')
```
Comment (crd1985) [inline framework/wazuh/tests/test_util.py]: I miss more examples. What about versions without the 'Wazuh ' prefix?

Comment (druizz90) [reply framework/wazuh/tests/test_util.py]: I added more unit tests:

```bash
# /var/ossec/framework/python/bin/pytest -s framework/wazuh/tests/test_util.py -vv
==================================================== test session starts ====================================================
platform linux -- Python 3.7.2, pytest-4.5.0, py-1.8.0, pluggy-0.11.0 -- /var/ossec/framework/python/bin/python3.7
cachedir: .pytest_cache
rootdir: /root/wazuh-old/wazuh/framework
collected 24 items                                                                                                          

framework/wazuh/tests/test_util.py::test_version_ok[Wazuh v3.5.0-Wazuh v3.5.2] PASSED
framework/wazuh/tests/test_util.py::test_version_ok[Wazuh v3.6.1-Wazuh v3.6.3] PASSED
framework/wazuh/tests/test_util.py::test_version_ok[Wazuh v3.7.2-Wazuh v3.8.0] PASSED
framework/wazuh/tests/test_util.py::test_version_ok[Wazuh v3.8.0-Wazuh v3.8.1] PASSED
framework/wazuh/tests/test_util.py::test_version_ok[Wazuh v3.9.0-Wazuh v3.9.2] PASSED
framework/wazuh/tests/test_util.py::test_version_ok[Wazuh v3.9.10-Wazuh v3.9.14] PASSED
framework/wazuh/tests/test_util.py::test_version_ok[Wazuh v3.10.1-Wazuh v3.10.10] PASSED
framework/wazuh/tests/test_util.py::test_version_ok[Wazuh v4.10.10-Wazuh v4.11.0] PASSED
framework/wazuh/tests/test_util.py::test_version_ok[Wazuh v5.1.15-Wazuh v5.2.0] PASSED
framework/wazuh/tests/test_util.py::test_version_ok[v3.6.0-v3.6.1] PASSED
framework/wazuh/tests/test_util.py::test_version_ok[v3.9.1-v3.9.2] PASSED
framework/wazuh/tests/test_util.py::test_version_ok[v4.0.0-v4.0.1] PASSED
framework/wazuh/tests/test_util.py::test_version_ok[3.6.0-3.6.1] PASSED
framework/wazuh/tests/test_util.py::test_version_ok[3.9.0-3.9.2] PASSED
framework/wazuh/tests/test_util.py::test_version_ok[4.0.0-4.0.1] PASSED
framework/wazuh/tests/test_util.py::test_version_ko[v3.6.0-v.3.6.1] PASSED
framework/wazuh/tests/test_util.py::test_version_ko[Wazuh v4-Wazuh v5] PASSED
framework/wazuh/tests/test_util.py::test_version_ko[Wazuh v3.9-Wazuh v3.10] PASSED
framework/wazuh/tests/test_util.py::test_version_ko[ABC v3.10.1-ABC v3.10.12] PASSED
framework/wazuh/tests/test_util.py::test_version_ko[Wazuhv3.9.0-Wazuhv3.9.2] PASSED
framework/wazuh/tests/test_util.py::test_version_ko[3.9-3.10] PASSED
framework/wazuh/tests/test_util.py::test_version_ko[3.9.0-3.10] PASSED
framework/wazuh/tests/test_util.py::test_version_ko[3.10-4.2] PASSED
framework/wazuh/tests/test_util.py::test_version_ko[3-3.9.1] PASSED
```

**Comment by druizz90** [comment]
## Issue update

#### Mocha tests

```javascript
# mocha test/test_agents.js --timeout=10000


  Agents
    GET/agents
      ✓ Request (378ms)
      ✓ Pagination (568ms)
      ✓ Retrieve all elements with limit=0 (523ms)
      ✓ Sort (438ms)
      ✓ Wrong Sort (473ms)
      ✓ Search (422ms)
      ✓ Selector (559ms)
      ✓ Not allowed selector (423ms)
      ✓ Version (524ms)
      ✓ Os.platform (500ms)
      ✓ Os.version (507ms)
      ✓ ManagerHost (533ms)
      ✓ Filters: status (542ms)
      ✓ Filters: status 2 (590ms)
      ✓ Filters: Invalid filter
      ✓ Filters: Invalid filter - Extra field
      ✓ Filters: older_than (450ms)
      ✓ Filters: group (450ms)
      ✓ Select: single field (462ms)
      ✓ Select: multiple fields (457ms)
      ✓ Select: wrong field (439ms)
      ✓ Select: invalid character
      ✓ Filters: query (482ms)
    GET/agents/summary
      ✓ Request (417ms)
    GET/agents/summary/os
      ✓ Request (391ms)
    GET/agents/outdated
      ✓ Request (509ms)
    GET/agents/:agent_id
      ✓ Request (manager) (446ms)
      ✓ Request (agent) (417ms)
      ✓ Selector (440ms)
      ✓ Not allowed selector (393ms)
      ✓ Params: Bad agent id
      ✓ Errors: No agent (459ms)
      ✓ Select (479ms)
      ✓ Select: wrong field (422ms)
    GET/agents/name/:agent_name
      ✓ Request (741ms)
      ✓ Wrong name (447ms)
      ✓ Selector (478ms)
      ✓ Not allowed selector (419ms)
    GET/agents/:agent_id/key
      ✓ Request (515ms)
      ✓ Params: Bad agent id
      ✓ Errors: No key (508ms)
    PUT/agents/groups/:group_id
      ✓ Request (412ms)
      ✓ Params: Bad group name
      ✓ Params: Group already exists (349ms)
    PUT/agents/:agent_id/group/:group_id
      ✓ Request (333ms)
      ✓ Params: Bad agent name
      ✓ Params: Agent does not exist (433ms)
      ✓ Params: Replace parameter (519ms)
    POST/agents/groups/:group_id/files/:file_name
      ✓ Request (895ms)
      ✓ ErrorOnBadGroup (409ms)
      ✓ ErrorOnEmptyConf
      ✓ OnlyAgentConfAllowed (375ms)
      ✓ InvalidConfDetected
      ✓ WrongConfDetected (552ms)
      1) TooBigXML
    GET/agents/no_group
      ✓ Request (462ms)
      ✓ Pagination (434ms)
      ✓ Retrieve all elements with limit=0 (435ms)
      ✓ Sort (423ms)
      ✓ Search (406ms)
      ✓ Select (338ms)
      ✓ Wrong select (387ms)
      ✓ Filter: status (451ms)
    GET/agents/groups
      ✓ Request (495ms)
      ✓ Retrieve all elements with limit=0 (423ms)
      ✓ Hash algorithm (366ms)
      ✓ Wrong Hash algorithm (423ms)
    GET/agents/groups/:group_id
      ✓ Request (645ms)
      ✓ Params: Bad group name (146ms)
      ✓ Retrieve all elements with limit=0 (440ms)
      ✓ Select (547ms)
      ✓ Filter: status (367ms)
    GET/agents/groups/:group_id/configuration
      ✓ Request (433ms)
      ✓ Params: Bad group name (691ms)
      ✓ Retrieve all elements with limit=0 (903ms)
    GET/agents/groups/:group_id/files
      ✓ Request (546ms)
      ✓ Params: Bad group name
      ✓ Retrieve all elements with limit=0 (508ms)
      ✓ Hash algorithm (525ms)
      ✓ Wrong Hash algorithm (567ms)
    GET/agents/groups/:group_id/files/:filename
      ✓ Request (484ms)
      ✓ UsingFormatAgentConfXML (460ms)
      ✓ UsingFormatAgentConfJSON (750ms)
      ✓ UsingFormatRootcheckXML (481ms)
      ✓ UsingFormatRootcheckJSON (403ms)
      ✓ Params: Bad group name
    POST/agents/groups/:group_id/configuration
      ✓ Request (548ms)
      ✓ ErrorOnBadGroup (394ms)
      ✓ ErrorOnEmptyConf
      ✓ InvalidConfDetected (118ms)
      ✓ WrongConfDetected (421ms)
      2) TooBigXML
    DELETE/agents/:agent_id/group
      ✓ Request (394ms)
      ✓ Errors: ID is not present (364ms)
      ✓ Params: Bad agent id
    DELETE/agents/:agent_id/group/:group_id
      ✓ Request (361ms)
      ✓ Errors: ID is not present (525ms)
      ✓ Errors: Group is not present (476ms)
      ✓ Params: Bad agent id
      ✓ Params: Bad group id (465ms)
    DELETE/agents/groups/:group_id
      ✓ Request (604ms)
      ✓ Params: Bad group id
    PUT/agents/restart
      ✓ Request (603ms)
    PUT/agents/:agent_id/restart
      ✓ Request (347ms)
      ✓ Params: Bad agent id
      ✓ Request (491ms)
    POST/agents/restart
      ✓ Request (496ms)
      ✓ Params: A good id and a bad one (347ms)
      ✓ Params: Bad agent id
      ✓ Request (403ms)
    DELETE/agents
      ✓ Request
      ✓ Filter: older_than, status and ids (3943ms)
      ✓ Errors: Get deleted agent (369ms)
      ✓ Filter: older_than (387ms)
    GET/agents/stats/distinct
      ✓ Request (324ms)
      ✓ Pagination (297ms)
      ✓ Retrieve all elements with limit=0 (286ms)
      ✓ Sort (311ms)
      ✓ Search (386ms)
      ✓ Select (324ms)
      ✓ Wrong select (347ms)
    GET/agents/:agent/config/:component/:configuration
      ✓ Request-Agent-Client (326ms)
      ✓ Request-Agent-Buffer (377ms)
      ✓ Request-Agent-Labels (370ms)
      ✓ Request-Agent-Internal (362ms)
      ✓ Request-Agentless-Agentless (359ms)
      ✓ Request-Analysis-Global (283ms)
      ✓ Request-Analysis-Active-response (328ms)
      ✓ Request-Analysis-Alerts (307ms)
      ✓ Request-Analysis-Command (277ms)
      ✓ Request-Analysis-Internal (279ms)
      ✓ Request-Auth-Auth (279ms)
      ✓ Request-Com-Active-response (303ms)
      ✓ Request-Com-Internal (281ms)
      ✓ Request-Csyslog-Csyslog (289ms)
      ✓ Request-Integrator-Integration (349ms)
      ✓ Request-Logcollector-Localfile (373ms)
      ✓ Request-Logcollector-Socket (339ms)
      ✓ Request-Logcollector-Internal (368ms)
      ✓ Request-Mail-Global (355ms)
      ✓ Request-Mail-Alerts (304ms)
      ✓ Request-Mail-Internal (337ms)
      ✓ Request-Monitor-Internal (348ms)
      ✓ Request-Request-Remote (326ms)
      ✓ Request-Request-Internal (443ms)
      ✓ Request-Syscheck-Syscheck (420ms)
      ✓ Request-Syscheck-Rootcheck (450ms)
      ✓ Request-Syscheck-Internal (448ms)
      ✓ Request-Wmodules-Wmodules (302ms)


  147 passing (1m)
  2 failing

  1) Agents
       POST/agents/groups/:group_id/files/:file_name
         TooBigXML:
     Uncaught AssertionError: expected '701' to be 701
      at Assertion.fail (node_modules/should/cjs/should.js:258:17)
      at Assertion.value (node_modules/should/cjs/should.js:335:19)
      at Test.<anonymous> (test/test_agents.js:1108:39)
      at Test.assert (node_modules/supertest/lib/test.js:179:6)
      at assert (node_modules/supertest/lib/test.js:131:12)
      at /root/wazuh-old/wazuh-api/node_modules/supertest/lib/test.js:128:5
      at Test.Request.callback (node_modules/superagent/lib/node/index.js:631:3)
      at /root/wazuh-old/wazuh-api/node_modules/superagent/lib/node/index.js:795:18
      at IncomingMessage.<anonymous> (node_modules/superagent/lib/node/parsers/json.js:16:7)
      at endReadableNT (_stream_readable.js:1139:12)
      at processTicksAndRejections (internal/process/task_queues.js:81:17)

  2) Agents
       POST/agents/groups/:group_id/configuration
         TooBigXML:
     Uncaught AssertionError: expected '701' to be 701
      at Assertion.fail (node_modules/should/cjs/should.js:258:17)
      at Assertion.value (node_modules/should/cjs/should.js:335:19)
      at Test.<anonymous> (test/test_agents.js:1843:39)
      at Test.assert (node_modules/supertest/lib/test.js:179:6)
      at assert (node_modules/supertest/lib/test.js:131:12)
      at /root/wazuh-old/wazuh-api/node_modules/supertest/lib/test.js:128:5
      at Test.Request.callback (node_modules/superagent/lib/node/index.js:631:3)
      at /root/wazuh-old/wazuh-api/node_modules/superagent/lib/node/index.js:795:18
      at IncomingMessage.<anonymous> (node_modules/superagent/lib/node/parsers/json.js:16:7)
      at endReadableNT (_stream_readable.js:1139:12)
      at processTicksAndRejections (internal/process/task_queues.js:81:17)

```

File: framework/wazuh/tests/test_util.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+import pytest
+import os
+
+from wazuh.utils import WazuhVersion
+
+test_data_path = os.path.join(os.path.dirname(os.path.realpath(__file__)), 'data')
+
+
+@pytest.mark.parametrize('version1, version2', [
```
Comment (crd1985) [inline framework/wazuh/tests/test_util.py]: I miss a use case. WazuhVersion('v3.9.0') < WazuhVersion('v3.9.0') should return a false but it is not being tested in this unit tests. Same for WazuhVersion('v3.9.0') > WazuhVersion('v3.9.0')

File: framework/wazuh/agent.py [hunk 3]
```diff
@@ -1889,11 +1889,18 @@ def get_outdated_agents(offset=0, limit=common.database_limit, sort=None, search
...
-                                      query="version<{};id!=0".format(manager.version) + ('' if not q else ';'+q),
-                                      get_data=True, count=True)
-        return db_query.run()
+                                      query=q, get_data=True, count=True)
+
+        list_agents_outdated = []
+        query_result = db_query.run()
+
+        for item in query_result['items']:
+            if WazuhVersion(item['version']) < WazuhVersion(manager.version):
```
Comment (crd1985) [inline framework/wazuh/agent.py]: If item['version'] is None or has a bad version format, an exception is raised. 
We should control that case properly.

File: framework/wazuh/tests/test_util.py (hunk 2 above)
Comment (druizz90) [reply framework/wazuh/tests/test_util.py]: Done!

File: framework/wazuh/agent.py (hunk 3 above)
Comment (druizz90) [reply framework/wazuh/agent.py]: Done!

**Comment by druizz90** [comment]
## Test performed

#### Unit tests

```bash
# /var/ossec/framework/python/bin/pytest framework/wazuh/tests/test_util.py -vv
================================================ test session starts ================================================
platform linux -- Python 3.7.2, pytest-4.5.0, py-1.8.0, pluggy-0.11.0 -- /var/ossec/framework/python/bin/python3.7
cachedir: .pytest_cache
rootdir: /shared/wazuh/framework
collected 28 items                                                                                                  

framework/wazuh/tests/test_util.py::test_version_ok[Wazuh v3.5.0-Wazuh v3.5.2] PASSED                         [  3%]
framework/wazuh/tests/test_util.py::test_version_ok[Wazuh v3.6.1-Wazuh v3.6.3] PASSED                         [  7%]
framework/wazuh/tests/test_util.py::test_version_ok[Wazuh v3.7.2-Wazuh v3.8.0] PASSED                         [ 10%]
framework/wazuh/tests/test_util.py::test_version_ok[Wazuh v3.8.0-Wazuh v3.8.1] PASSED                         [ 14%]
framework/wazuh/tests/test_util.py::test_version_ok[Wazuh v3.9.0-Wazuh v3.9.2] PASSED                         [ 17%]
framework/wazuh/tests/test_util.py::test_version_ok[Wazuh v3.9.10-Wazuh v3.9.14] PASSED                       [ 21%]
framework/wazuh/tests/test_util.py::test_version_ok[Wazuh v3.10.1-Wazuh v3.10.10] PASSED                      [ 25%]
framework/wazuh/tests/test_util.py::test_version_ok[Wazuh v4.10.10-Wazuh v4.11.0] PASSED                      [ 28%]
framework/wazuh/tests/test_util.py::test_version_ok[Wazuh v5.1.15-Wazuh v5.2.0] PASSED                        [ 32%]
framework/wazuh/tests/test_util.py::test_version_ok[v3.6.0-v3.6.1] PASSED                                     [ 35%]
framework/wazuh/tests/test_util.py::test_version_ok[v3.9.1-v3.9.2] PASSED                                     [ 39%]
framework/wazuh/tests/test_util.py::test_version_ok[v4.0.0-v4.0.1] PASSED                                     [ 42%]
framework/wazuh/tests/test_util.py::test_version_ok[3.6.0-3.6.1] PASSED                                       [ 46%]
framework/wazuh/tests/test_util.py::test_version_ok[3.9.0-3.9.2] PASSED                                       [ 50%]
framework/wazuh/tests/test_util.py::test_version_ok[4.0.0-4.0.1] PASSED                                       [ 53%]
framework/wazuh/tests/test_util.py::test_version_ko[v3.6.0-v.3.6.1] PASSED                                    [ 57%]
framework/wazuh/tests/test_util.py::test_version_ko[Wazuh v4-Wazuh v5] PASSED                                 [ 60%]
framework/wazuh/tests/test_util.py::test_version_ko[Wazuh v3.9-Wazuh v3.10] PASSED                            [ 64%]
framework/wazuh/tests/test_util.py::test_version_ko[ABC v3.10.1-ABC v3.10.12] PASSED                          [ 67%]
framework/wazuh/tests/test_util.py::test_version_ko[Wazuhv3.9.0-Wazuhv3.9.2] PASSED                           [ 71%]
framework/wazuh/tests/test_util.py::test_version_ko[3.9-3.10] PASSED                                          [ 75%]
framework/wazuh/tests/test_util.py::test_version_ko[3.9.0-3.10] PASSED                                        [ 78%]
framework/wazuh/tests/test_util.py::test_version_ko[3.10-4.2] PASSED                                          [ 82%]
framework/wazuh/tests/test_util.py::test_version_ko[3-3.9.1] PASSED                                           [ 85%]
framework/wazuh/tests/test_util.py::test_same_version[Wazuh v3.10.10-Wazuh v3.10.10] PASSED                   [ 89%]
framework/wazuh/tests/test_util.py::test_same_version[Wazuh v5.1.15-Wazuh v5.1.15] PASSED                     [ 92%]
framework/wazuh/tests/test_util.py::test_same_version[v3.6.0-v3.6.0] PASSED                                   [ 96%]
framework/wazuh/tests/test_util.py::test_same_version[v3.9.2-v3.9.2] PASSED                                   [100%]

============================================= 28 passed in 0.17 seconds =============================================

```

**Pull request merged**